### PR TITLE
Supercruise Part 2 - Randomly generated ruins, Infinite shuttle Zs, Economy changes, Partial science refactor, Exploration job.

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## This PR has a fucked up merge conflict resolution so will be closed in favour of an identical PR that isn't conflicting when I finish porting the map changes and icons from this PR to the new one.

Dependant on https://github.com/BeeStation/BeeStation-Hornet/pull/4390

**Any map ruin parts must be of dimensions 4a+1 by 4b+1, where a and b are positive non-zero integers.**

### TODO
 - [x] Ruin Generator
 - [ ] Add in more ruin parts **(ONGOING)**
 - [x] Add in mission events - Go to a ruin to do a specific job
 - [x] Add in recover blackbox mission
 - [ ] Add in kill enemy mission
 - [x] Nuclear Decomission mission
 - [x] Remove budget cards being given money randomly.
 - [x] Money is distributed to the station upon mission completion.
 - [x] Add in 'unique' elements to stations (Can't have 2 armourys etc.)
 - [x] Add in weighting to room spawning.
 - [x] Are rooms and halls spawning from 1 spawn in some instances?
 - [x] Improve the mission code so that missions are viewable from a command console and multiple are assigned at once.
 - [x] Prevent atmos from venting stations during generation.
 - [x] Add in airlocks between rooms.
 - [ ] Add in firedoors between hallways.
 - [x] Map objective consoles.
 - [x] Ghost role level support.
 - [ ] <s>Pre-existing ruin levels.</s>
 - [x] Better loot (New research nodes that can only be unlocked via getting the disks at the ruins)
 - [x] Wiping z-levels doesn't account for being in lockers
 - [x] Ships shouldn't be able to dock on plating
 - [ ] Emergency extraction fulton returns you to station if you get zcleared.
 - [x] Zclearing dead mobs should store them and then put them on the next generated ruin.
 - [x] Ships need a thruster buff
 - [x] VIP needs to spawn with tools
 - [ ] When nobody on a shuttle can pilot, it should automatically dethrottle
 - [x] Spiders need to be nerfed
 - [x] ZClear needs to be blocked by shuttles
 - [ ] Ferry shuttle needs cheating autopilot
 - [ ] Exploration shuttle?
 - [x] Remove the admin message warning
 - [ ] <s>Give curator some kind of shuttle</s>
 - [x] Shuttles can dethrottle and dock at the same time, bugging it out
 - [ ] Corg transport shuttle shouldn't be able to fly away from the station
 - [ ] When remote controlled shuttles dock, they get transported to the remote control dock even if they dock at a different z-level
 - [ ] Add in computer to communicate with curator
 - [ ] Curator has a book with solutions to some puzzles

### Ideas

New jobs dedicated to exploration:
 - Exploration shuttle
 - Exploration crew
 - Off-station firing pins
 - Curator somehow ties into this. (Reading relics that nobody else can read?)
 - Get all old space ruin loot and stick it on the stations so there is at least some loot

Exploration radio channel, servers are setup inside the exploration shuttle.

image

### Mapping Changes (So many map files that diff bot is broken)

Corgstation:
image
image

 - Removed the transit tube to the asteroid
 - Asteroid now connects to the main maints
 - Northeast maints now connects to the sec/cargo hallway so it is not completely separate.
 - Mining moved to the south
 - Exploration bay added south
 - Dorms moved north
 - Warehouse moved north.
 - Maints changed

## About The Pull Request

Implements randomly generated ruin levels into supercruise.
Implements dynamic z-levels, which will be reset upon no longer being needed.
Implements an advanced ruin generation system.
Implements ZClear subsystem.

image

Ruin signals are no longer linked to a z-level on the map. Instead when a shuttle attempts docking with them, the ruin is assigned an empty z-level from ZClear and a random ruin is generated.
By the time the shuttle pilot docks with the z-level the ruin will have generated.
When the shuttle leaves the z-level and no more conscious mobs are in it, z-clear will automatically clean up the z-level.

Floating out from the station will put you into the randomly generated z-levels, however nothing will be generated. 

Ruin generation works by taking a bunch of map files with markers for rooms and halls. It places rooms and halls in a manner such that every single marker for room and halls will always be used up. Once room generation is completed, it will decorate the floors with random stuff.

Once all mobs leave a ruin z-level then the beacon is deleted and a new one will be generated again some time later.

If the amount of required dynamic z-levels exceeds the number of existing dynamic z-levels a new one will be created.

Currently there are 4 types of decorations that are applied:
Abandoned: Just a bunch of trash everywhere.
Blob: A bunch of trash + blob mobs.
Xeno: A bunch of trash, broken lights, blood and xenomorphs.
Netherworld: Cult pylons, nethermobs etc.

image

image

Abandoned:

image

image

Netherworld

image

Xeno

image

Clockcult

image

### Objectives

Blackbox Recovery:
Difficulty: Pretty Easy
Contact with the outpost has been lost and Nanotrasen need answers and they need them fast! Enter the outpost, locate the blackbox and return it to the main station's bridge to complete the objective. Pays out between 50000 and 100000 credits between departments.

Nuclear Decomissioning
Difficulty: Average
The press can't get a hold of what happened on that outpost, a nuclear bomb will be delivered to the bridge and the outdated nuclear authentication disk must be located on the ruin. Once inserted into the nuclear bomb, the station will be destroyed granting a payout of between 80000 and 200000 credits.

VIP Recovery (WIP)
Difficulty: Hard
A VIP has been located on the outpost and must be recovered. The VIP will either spawn as a dead body with a diary, or controlled by a ghost player. If controlled by a ghost player, the mission will fail when they die (On a station filled with enemies, what can go wrong?). VIP can potentially be hostile towards the people trying to recover them and must be non-lethally restrained. On top of that, they do not have a GPS beacon so have to be located.
Pays out between 100000 to 250000 credits.

### Research Changes

At the moment research is a matter of building a bunch of research servers, and finishing the entire techweb in 20 minutes. Some research nodes that provide powerful items that aren't required for the function of the station have been moved to research nodes that can be discovered on derelict stations.

Derelict located research:
 - Bags of holding
 - Portal Gun
 - Quantum spin inverter
 - Combat Implants
 - Armblade implant
 - Nuclear energy gun
 - Xray gun
 - Beam rifle
 - Exotic Ammunition
 - Phazon
 - Clusterbang Launcher

Research is relatively common, spawning in certain rooms, spawning in most derelict loot rooms (Derelicts have 1 marked loot room) and spawning with a 0.5% chance on every single open floor tile.

### Economy Changes

The economy has been significantly rebalanced. Station budget cards will no longer randomly be injected with funds every 5 minutes. Instead to gain station funds, bounties or station objectives must be completed. Bounties now pay off significantly more, however will be distributed into all departments budgets.

Additionally, the wages have been changed.
MINIMUM 5 -> 10
EASY 20 -> 15
MEDIUM 50 -> 40
HARD 100 -> 70
Command 200 -> 100

For reference
Assistants/Mimes/Clowns used to start with 25 credits, and will now start with 50 credits.
Bartender/Botanists/Cargo/Chaplain/Cook/Curator/Deputy/Gimmick/Janitor/Lawyers used to start with 200 credits and will now start with 75.
Atmos, Briggie, Chemist, Detective, EMT, Genetisist, MD, QM, Robotics, Science, Enginees and Virologists used to start with 250 credits and will now start with 200 credits.
Sec/Miners/Warden used to start with 500 credits and will now start with 350.
Command used to start with 1000 credits and will now start with 500.

Overall wages have been reduced and made closer together so that budget cards wont be drained so quickly and prices can be more fair.

Each department has been changed so that it now starts with 10000 credits as opposed to 7000.

Paychecks will take out ~300 to ~500 credits every 5 minutes, and missions reward upwards of (5K/10K depending on department) meaning that 1 mission should support a heavy department for >1 hour, however of course extra funds are needed for buying things.

Instead of money going directly to cargo for them to buy random shit, it will now be distributed to the heads where they can request things.

Some of the prices of non-department related vending machines have been reduced.

## Why It's Good For The Game

Random ruins, something fun and unique every single time.

## Changelog
:cl:
add: Adds in randomly generated ruins
add: Adds in dynamic z-levels.
add: Adds in ruin beacons on the shuttle supercruise map.
code: Adds z-clear
code: Ruin mapsize validation unit test
code: Fixes fastdmm2 not compiling on beestation code due to the updates.
refactor: Budget cards no longer get free money.
refactor: Bounties reward 3x as much, but the reward is distributed among station budgets.
refactor: Paychecks have been slightly reduced.
refactor: Station budgets now start with more money to compensate early game when no objectives or bounties have been completed.
refactor: Research disks can be found on ruin stations. These can be used to unlock special research.
refactor: Bags of holding are now a separate research node unlocked by finding disks on derelict stations.
refactor: Wormhole projectors are now a separate research node unlocked by finding disks on derelict stations.
refactor: Quantum spin inverters are now a separate research node unlocked by finding disks on derelict stations.
refactor: Advanced combat cyber implants are now a separate research node unlocked by finding disks on derelict stations. (Hydraulic armblades)
refactor: Combat cybernetic implants implants are now a separate research node unlocked by finding disks on derelict stations. (Xray, thermal vision, anti stun implants, thrust implants)
refactor: Clusterbangs are now a found research disk.
refactor: Phazons are now a found research disk.
refactor: Exotic ammo is now a found research disk.
refactor: Beam rifles are now a found research disk.
refactor: Temp gun and X-ray gun is now a found research disk.
refactor: Nuclear energy gun is now a found research disk.
add: Completion station missions reward money to budget cards.
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
> Any map ruin parts must be of dimensions 4a+1 by 4b+1, where a and b are positive non-zero integers.

This needs to be enforced by CI or a unit test, IMO.

**Comment by ZombiesVsPlants** [comment]
This is so fucking cool

**Comment by OceanFish1** [comment]
I just hope navigation computers won't give you TGUI's BSOD again.
I still like the idea though

**Comment by Misha600** [comment]
i love you

**Comment by PowerfulBacon** [comment]
Wrong button

**Comment by RDS88-beeman** [comment]
What if greytide decor

**Comment by PowerfulBacon** [comment]
Gunna testmerge this over part 1, since then I can test both of the PRs at the same time, and this is gameplay ready.

**Comment by snek-with-tophat** [comment]
Does this remove the ghost spawners from space?
What about old space ruins?

**Comment by PowerfulBacon** [comment]
> 
> 
> Does this remove the ghost spawners from space?
> What about old space ruins?

the only there is is the derelict drone, which is super boring and only used to go to the station which they aren't meant to do

**Comment by snek-with-tophat** [comment]
What about hotel staff ,charlie crew and the lone comms agent?

**Comment by PowerfulBacon** [comment]
> 
> 
> What about hotel staff ,charlie crew and the lone comms agent?

oh yea those too, yea they are gone but I can implement random station levels having ghost roles.

**Comment by ghost** [comment]
Could I ask some questions?
1. are the ruins just in forms of derelict stations?
2. will there be plans for ghost roles in the ruins?
3. Is there multiple ruins per round?

**Comment by PowerfulBacon** [comment]
> 
> 
> Could I ask some questions?
> 
>     1. are the ruins just in forms of derelict stations?

All ruins take the form of a randomly generated station with some kind of disaster applied.

>     2. will there be plans for ghost roles in the ruins?

Potentially, they are very easy to add but should be capped to a max of 1 ghost role ruin since ghost roles block the freeing of the z-level.

>     3. Is there multiple ruins per round?

There are infinite.

**Comment by ghost** [comment]
Is there any asteroids or syndicate base space ruins planned?

**Comment by PowerfulBacon** [comment]
> 
> 
> Is there any asteroids or syndicate base space ruins planned?



> 
> 
> Is there any asteroids or syndicate base space ruins planned?

May incorporate some beacons which have the existing space ruins, however all ghost role ruins will be replaced with ghost role spawns on the generated ruins.

**Comment by PowerfulBacon** [comment]
Ships in the process of docking should block zclear

**Comment by PowerfulBacon** [comment]
them spiders are deadly

**Comment by PowerfulBacon** [comment]
VIP needs tools

**Comment by PowerfulBacon** [comment]
Wiping z-levels doesnt account for being inside lockers, and the zclear is wiping bodies.

**Comment by PowerfulBacon** [comment]
Shuttles should dethrottle when eveyone on them dies

**Comment by PowerfulBacon** [comment]
Disconnected players should count as mentally active

File: code/controllers/subsystem/zclear.dm [hunk 1]
```diff
@@ -34,11 +34,24 @@ SUBSYSTEM_DEF(zclear)
 */
 /datum/controller/subsystem/zclear/proc/check_for_empty_levels()
 	var/list/active_levels = list()
-	for(var/mob/living/L in GLOB.player_list)
-		active_levels["[L.z]"] = TRUE
+	//Check active mobs
+	for(var/mob/living/L in GLOB.mob_list)
```
Comment (ike709) [inline code/controllers/subsystem/zclear.dm:44]: Do a typeless check with living_mob_list

**Comment by Absolutely-Steph** [comment]
This is amazing, you could run a whole server off of just this mechanic, why doesn't NSV do it like this.

**Comment by PowerfulBacon** [comment]
A grace period is needed when leaving and entering z-levels

**Comment by PowerfulBacon** [comment]
Needs ghost roles, needs easier enemies, the beacon needs to be deleted before wiping the z-level so you cant go to a wiping z-level.

**Comment by PowerfulBacon** [comment]
It appears I have majorly fucked up and turned 1 merge conflict into thousands

**Comment by PowerfulBacon** [comment]
I dont know how to unfuck this

**Comment by Boxlta** [comment]
Few questions.  For the loot, what sorts of loot will be available? I see a few stechkins, and a ruined armory on the xeno one. Will the armory lockers have normal armory locker spawns? Also, with the xeno one, I see xeno eggs. Will these be able to start a xenohive player controlled if brought back to station?

**Comment by Mat05usz** [comment]
Question: Could this be used to make fully randomized maint? Not just rooms but also 'hallways' between them and placement?

**Comment by PowerfulBacon** [comment]
Closed in favour of https://github.com/BeeStation/BeeStation-Hornet/pull/4543

**Pull request closed without merging**